File: 백준/Silver/2667. 단지번호붙이기/단지번호붙이기.py
from collections import deque

def bfs(start_x, start_y, graph, visited):
    # 상, 하, 좌, 우 이동을 위한 방향 벡터
    dx = [-1, 1, 0, 0]
    dy = [0, 0, -1, 1]
    count = 1  # 시작 지점도 단지에 포함되므로 1로 시작
    queue = deque([(start_x, start_y)])
    visited.add((start_x, start_y))  # 시작 좌표를 방문처리
    
    while queue:
        x, y = queue.popleft()
        # 4방향 탐색
        for i in range(4):
            nx = x + dx[i]
            ny = y + dy[i]
            if 0 <= nx < len(graph) and 0 <= ny < len(graph[0]):
                if (nx, ny) not in visited and graph[nx][ny] == '1':
                    visited.add((nx, ny))
                    queue.append((nx, ny))
                    count += 1  # 단지에 포함되는 집의 수를 증가
    
    return count

# 입력 처리
a = int(input())
graph = [list(input().strip()) for _ in range(a)]

visited = set()  # 방문한 좌표 저장
nums = []  # 단지 크기 저장

# 전체 좌표를 순회하며 단지를 찾음
for i in range(a):
    for j in range(a):
        if graph[i][j] == '1' and (i, j) not in visited:
            # 새로운 단지를 찾으면 BFS를 실행
            nums.append(bfs(i, j, graph, visited))

# 결과 출력
nums.sort()  # 단지 크기를 오름차순 정렬
print(len(nums))  # 총 단지 수
for num in nums:
    print(num)  # 각 단지의 크기 출력

File: 백준/Silver/2667. 단지번호붙이기/단지번호붙이기.java
import java.io.BufferedReader;
import java.io.IOError;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Collections;
import java.util.LinkedList;
import java.util.List;
import java.util.Queue;
import java.util.StringTokenizer;

public class Main {
    public static int dfs(int start_x, int start_y ,int[][] graph, int[][]visited){
        visited[start_x][start_y]=1;
        int count=1;
        int [] dx ={-1,1,0,0};
        int[] dy={0,0,-1,1};

        for(int i=0; i<4;i++){
            int nx=dx[i]+start_x;
            int ny=dy[i]+start_y;
            if(nx>=0 && nx<graph.length && ny>=0 && ny<graph[0].length && visited[nx][ny]!=1 && graph[nx][ny]==1){
                count+=dfs(nx,ny,graph, visited);

            }
        }
        return count;

    }


    public static void main(String[] args) throws IOException {
        BufferedReader bf = new BufferedReader(new InputStreamReader(System.in));
        int n= Integer.parseInt(bf.readLine());
        int num=0;
        int[][] graph = new int[n][n];
        int[][] visited= new int[n][n];
        List<Integer>result = new LinkedList<>();

        for (int i=0; i<n; i++){
            String line = bf.readLine();
            for(int j=0; j<n; j++){
                graph[i][j] = line.charAt(j)-'0';
            }
        }
        for (int x=0; x<n; x++){
            for(int y=0; y<n; y++){
                if(graph[x][y]==1& visited[x][y]!=1){
                    result.add(dfs(x,y,graph,visited));
                    num+=1;
                }
            }
        }
        System.out.println(num);
        Collections.sort(result);
        for (int j=0; j<result.size(); j++){
            System.out.println(result.get(j));
        }






    }
}